### PR TITLE
[diffusion, trainer] refactor: project TQ metric reads

### PR DESCRIPTION
**Pull request by dongbo910220**
### What does this PR do?

Reduce diffusion V1 metric-read overhead by retrieving only the TransferQueue fields consumed by metric computation. Training, validation, and rollout dumps keep full-payload reads, while lightweight response-shape metadata preserves image-count logging without fetching generated images.

No open duplicate was found in this [search](https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+is%3Aopen+%22TransferQueue+select_fields+diffusion%22). 

### Test

```bash
CUDA_VISIBLE_DEVICES='' python -m pytest -q \
  tests/trainer/diffusion/test_v1_tq_field_projection_on_cpu.py \
  tests/agent_loop/test_diffusion_agent_loop_correctness_on_cpu.py \
  tests/trainer/diffusion/test_v1_direct_preference_on_cpu.py \
  tests/trainer/diffusion/test_diffusion_metric_utils_on_cpu.py \
  tests/trainer/diffusion/test_v1_replay_buffer_on_cpu.py
```

Result: `all passed` 

Measured with real pretrained SD3.5 Medium on 2× RTX 5090 D using two complementary 36-step V1 synchronous TransferQueue runs (four warmup steps per run):

| Metric | main | PR | Improvement |
| --- | ---: | ---: | ---: |
| E2E step | 12.8055 s | 12.4885 s | 2.48% less time |
| `_compute_metrics` | 147.50 ms | 49.66 ms | 66.33% less time |

Raw E2E is the observed end-to-end point estimate; `_compute_metrics` is the directly localized result.

### Design & Code Changes

- Add optional TransferQueue field projection to diffusion batch conversion and use it only for V1 metric reads.
- Carry response shape in rollout tags for metric logging without reading the response tensor.
- Add focused correctness tests and a real-TransferQueue benchmark.

### CODE REVIEW

File: scripts/benchmark_diffusion_tq_field_projection.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        "sample_level_rewards",
+        "sample_level_scores",
+        "advantages",
+        "returns",
+        "uid",
+        "extra_fields",
+        "sum_pi_squared",
+        "old_log_probs",
+        "response_mask",
+        "rollout_is_weights",
```
[the remark this answers is not in the document]
Comment (dongbo910220) [reply scripts/benchmark_diffusion_tq_field_projection.py]: Fixed in `dbd1e13`. The benchmark now selects fields by `--algorithm` through the same `diffusion_metric_tq_fields` helper used by the trainer. Policy-gradient and direct-preference now request only fields present in their payloads; both real TransferQueue 0.1.8 runs completed successfully.

File: verl_omni/trainer/diffusion/v1/trainer_base.py [hunk 2]
```diff
@@ -1499,7 +1500,22 @@ def _val_metrics_update(self, data_sources, sample_uids, reward_extra_infos_dict
...
+                "sample_level_rewards",
+                "sample_level_scores",
+                "advantages",
+                "returns",
+                "uid",
+                "extra_fields",
+                "sum_pi_squared",
+                "old_log_probs",
+                "response_mask",
+                "rollout_is_weights",
```
[the remark this answers is not in the document]
Comment (dongbo910220) [reply verl_omni/trainer/diffusion/v1/trainer_base.py]: Fixed in `dbd1e13`. `_compute_metrics` now uses an algorithm-specific projection: policy-gradient requests rewards, scores, advantages, returns, `uid`, and extra fields, while direct-preference requests only rewards, scores, `uid`, and extra fields. The absent optional fields are no longer selected. Focused tests: all passed.

**Comment by zhtmike** [comment]
Ping @knlnguyen1802 for this.

File: verl_omni/trainer/diffusion/v1/tq_utils.py [hunk 3]
```diff
@@ -28,6 +28,17 @@
 logger.setLevel(os.getenv("VERL_LOGGING_LEVEL", "INFO"))
 
 
+def diffusion_metric_tq_fields(
+    algorithm: Literal["policy_gradient", "direct_preference"],
+) -> list[str]:
+    """Return fields persisted and consumed by diffusion metric computation."""
+    if algorithm == "policy_gradient":
+        return ["sample_level_rewards", "sample_level_scores", "advantages", "returns", "uid", "extra_fields"]
```
Comment (knlnguyen1802) [inline verl_omni/trainer/diffusion/v1/tq_utils.py:47]: **Suggestion:** Lock this list against the persist writes in `_train_policy_gradient_batch` / `_train_direct_preference_batch`. The first head requested `sum_pi_squared`, `response_mask`, and `rollout_is_weights`, which TransferQueue rejects when they are not stored. A CPU test that `diffusion_metric_tq_fields(algo)` is a subset of the corresponding `put_dataproto_fields_to_tq(..., fields=...)` list plus the rollout-written `uid` / `extra_fields` would keep that from drifting again.

File: verl_omni/trainer/diffusion/v1/trainer_base.py [hunk 4]
```diff
@@ -1508,16 +1516,42 @@ def _compute_metrics(self, batch_meta: KVBatchMeta, metrics, timing_raw, global_
             if "advantages" in data.batch
             else data.batch["sample_level_scores"].shape[0]
         )
-        responses = data.batch.get("responses")
-        real_images = 0
-        if isinstance(responses, torch.Tensor) and responses.numel() > 0 and responses.dim() >= 4:
-            real_images = int(responses.shape[0])
+        response_shapes = []
+        for tag in batch_meta.tags:
```
Comment (knlnguyen1802) [inline verl_omni/trainer/diffusion/v1/trainer_base.py]: **Suggestion:** Skip `is_padding` tags here the same way the staleness block below already does. One padding or historical tag without `response_shape` currently discards the whole batch and warns every step, even when the real trajectories have the tag.

File: tests/trainer/diffusion/test_v1_tq_field_projection_on_cpu.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+    def get_fields(**kwargs):
+        captured.update(kwargs)
+        return {
+            "sample_level_rewards": torch.ones(2, 3),
+            "uid": ["first", "second"],
+        }
+
+    monkeypatch.setattr(tq_utils.tq, "kv_batch_get", get_fields)
+    selected = ["sample_level_rewards", "uid"]
```
Comment (knlnguyen1802) [inline tests/trainer/diffusion/test_v1_tq_field_projection_on_cpu.py]: **Suggestion:** This conversion test never selects `extra_fields`, and the `_compute_metrics` tests inject already-unpacked `reward_extra_info`. The production path depends on projected `extra_fields` being unpacked into `reward_extra_info` for `critic/*/mean`. Please add one conversion case that selects `extra_fields` and asserts the unpacked extra-info keys survive.

**Review by knlnguyen1802 — COMMENTED** [review]
See inline suggestions.

File: verl_omni/trainer/diffusion/v1/tq_utils.py (hunk 3 above)
Comment (dongbo910220) [reply verl_omni/trainer/diffusion/v1/tq_utils.py:47]: Both persistence paths now use `diffusion_persisted_tq_fields()`, and a parameterized CPU contract test verifies that each metric projection is a subset of those persisted fields plus the rollout-written `uid` and `extra_fields`.

File: verl_omni/trainer/diffusion/v1/trainer_base.py (hunk 4 above)
Comment (dongbo910220) [reply verl_omni/trainer/diffusion/v1/trainer_base.py]: Response-shape telemetry now filters `is_padding` tags before validation and aggregation, so padding without shape metadata no longer invalidates real trajectories or contributes to the real-image count. Added a mixed real/padding regression case.

File: tests/trainer/diffusion/test_v1_tq_field_projection_on_cpu.py (hunk 5 above)
Comment (dongbo910220) [reply tests/trainer/diffusion/test_v1_tq_field_projection_on_cpu.py]: The conversion test now selects `extra_fields` and verifies that the projected `reward_extra_info` dictionaries are unpacked intact into `non_tensor_batch`, covering the production input used for critic metrics.

**Review by knlnguyen1802 — APPROVED** [review]
Thanks for the optimization

**Comment by NancyFyong** [comment]
pls fix ci error

**Comment by wtomin** [comment]
@dongbo910220  Please fix cpu unit test error. The smoke test error is occasional. You can leave it alone.

**Comment by dongbo910220** [comment]
> @dongbo910220 Please fix cpu unit test error. The smoke test error is occasional. You can leave it alone.

OK, I have updated commit.

**Comment by dongbo910220** [comment]
> pls fix ci error

fixed.

**Comment by dongbo910220** [comment]
Hi @NancyFyong, I have merged the latest `main` into this PR, including #603, and rerun the full CPU unit test suite locally: all passed. Could you please add `ready-for-ci` to run CI again? Thanks!

**Comment by dongbo910220** [comment]
Hi @NancyFyong, the latest `gpu_smoke` failure does not appear to be caused by this PR. The same failure is reproducible on `main`: six diffusion E2E cases and one omni E2E case fail during LoRA weight synchronization because `collect_lora_params` returns no parameters ([main run](https://github.com/verl-project/verl-omni/actions/runs/35080383806)). [#607](https://github.com/verl-project/verl-omni/pull/607) addresses this issue, and its GPU smoke has passed.
